Show Engine Version on the UI + Consolidate Engine Version and Game Configuration

DanVanAtta opened the pull request:
### Changes:
- Parse the game engine version from a config file instead of hardcoded
- Convert the mapDownload.properties file into a generic config file 'game_engine.properties'
- Display the game engine version on the UI
- Update Travis to read and update the game version in 'game_engine.properties'

### Before / Context:
-  `EngineVersion.java` stored the game engine version in statically initialized singleton instance
- Travis also stored the game engine version redundantly as an environment variable (this can be removed after this PR)
- MapDownloadController.java reads from the 'mapDownload.properties' to know where the triplea_maps.xml file is

### General Approach
- Have just one configuration file rather than many: 'game_engine.properties'
- Have the EngineVersion parse the file and ignore the fifth digit, yet still store it so we can display it on the UI.
- Push the game engine config parsing code 'up' so that it is not duplicated by anything that loads a value from config.
- The code to parse the property file needs a home. It is consolidated into a `PropertyReader` implementation, which is then needed to use an `EngineVersion` or a `MapDownloadController`. The next key problem is that `EngineVersion` is used in 30+ places, so wiring that data through to use EngineVersion each time is not very practical. So to solve this we consolidate the initialization code to a `ClientContext` object. Then the old code that used to get an EngineVersion via singleton, it gets it through the ClientContext singleton now instead which can return an initialized EngineVersion.
- ClientContext is modeled after Spring IOC containers. In this case it helps since we can conslidate initialization logic, and not pass around objects needed for initialization in 31+ places. (See bottom of this summary for an architecture diagram, showing these transformations/reorganization)


### Comments on the  IOC container `ClientContext`:
- `ClientContext` is reinventing the wheel..  But for now it helps a lot, and we don't have to go through the hurdle yet of picking and using a full blown IOC container.
- The test infrastructure for the IOC is not complete. There for now is no way to prevent normal IOC container initialization when running tests. A hook has been left in so at least IOC container objects can be mocked, but to really go the distance we would want to stop the normal initialization completely.

### Circular Dependency Note When reading Game engine version from Config file
- The game engine version was moved to a config file, at the root folder
- To find the root folder, we need the engine version
- To read the config file, to get the engine versoin, we need the root folder..

To break this circular loop it wasn't too hard to drop the need for the engine version when finding the root folder. A regex could match any version number, and there is a specific prefix and suffix that we were looking for before, so the actual version number was not important. By removing that we no longer had the circular call loop.

### Methods moved from GameRunner2 to ClientFileSystemHelper
- While trying to resolve the circular call loop I moved stuff out of GameRunner2 to ClientContext. Then eventually this stuff moved to `ClientFileSystemHelper`. The goal there is to place a lot of the procedural utility code somewhere that is not GameRunner2 and not ClientContext. 

### UI  Screenshots
Before:
![before](https://cloud.githubusercontent.com/assets/12397753/12775315/daa40136-ca01-11e5-9bae-e18ca8994970.png)

After with game engine version:
![after](https://cloud.githubusercontent.com/assets/12397753/12775316/daa4249a-ca01-11e5-847c-7e3661e276c0.png)

Final with renamed "game name" and "game version" labels:
![after_after](https://cloud.githubusercontent.com/assets/12397753/12775380/3cf6e47a-ca02-11e5-8b3d-bb09885e2172.png)


### High level Change Diagram

Goal: share the data pulled from the config file between MapDownloadController and EngineVersion. To do this they will share the `congFileParser` , and a ClientContext object will create the common dependencies and then instantiate objects that need them. 

![arch_diagram_for_ioc](https://cloud.githubusercontent.com/assets/12397753/12776663/c828d828-ca0c-11e5-90f4-c94f3176b9db.png)

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/triplea-game/triplea/426)
<!-- Reviewable:end -->

DanVanAtta commented on .gitattributes:
```diff
@@ -1,11 +1,13 @@
 # Set the default behavior, in case people don't have core.autocrlf set.
 * text=auto
 
+MacOS_users_read_this_first.txt eol=lf
```
Moving the exceptions towards the top, so they stand out more (I was hoping order would matter and rules would take precedence, not the case)

gaborbernat replied:
I like the approach, will review it in depth sometimes when I have the free
time for it.

Bernát GÁBOR

On Wed, Feb 3, 2016 at 9:34 AM, Dan Van Atta <notifications@github.com>
wrote:

> In .gitattributes
> <https://github.com/triplea-game/triplea/pull/426#discussion_r51691525>:
>
> > @@ -1,11 +1,13 @@
> >  # Set the default behavior, in case people don't have core.autocrlf set.
> >  * text=auto
> >
> > +MacOS_users_read_this_first.txt eol=lf
>
> Moving the exceptions towards the top, so they stand out more (I was
> hoping order would matter and rules would take precedence, not the case)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/triplea-game/triplea/pull/426/files#r51691525>.
>

DanVanAtta commented on .gitattributes:
```diff
@@ -15,8 +17,7 @@
 *.plist eol=lf
-MacOS_users_read_this_first.txt eol=lf
-triplea_maps.xml eol=lf
+*.properties eol=lf
```
CRLF endings was giving a lot of trouble for scripting. Doesn't make sense to force "^M" characters to then read it back and deal with stripping it away.

DanVanAtta commented on .travis.yml:
```diff
@@ -7,7 +7,10 @@ script:
 before_deploy:
 - wget -O setup https://github.com/triplea-game/triplea/blob/master/.travis/setup?raw=true
 - chmod +x setup && ./setup
-- export TAGGED_VERSION="$BUILD_VERSION.$TRAVIS_BUILD_NUMBER"
```
Note that 'BUILD_VERSION' is no longer used, this is a redundant place where we had the game engine version. Otherwise below we read from "game_engine.properties", parse out the version, set that to "ENGINE_VERSION", then we append the build number, and write it back out to the game_engine.properties file.

DanVanAtta commented on build.gradle:
```diff
@@ -114,7 +114,7 @@ task allPlatform(type: Zip, group: 'release', dependsOn: [renameShadowJar, updat
 		into('icons')
 		exclude("icon.icns")
 	}
-	['readme.html', 'changelog.txt', 'system.ini', 'TripleA_RuleBook.pdf', 'mapDownload.properties',
```
Renaming 'mapDownload.properties' to 'game_engine.properties', and consolidating game config to the one file.

DanVanAtta commented on game_engine.properties:
```diff
@@ -1,3 +1,4 @@
+engine_version = 1.8.0.10
```
new engine_version property being added here - this is the one location where we will have this value!

DanVanAtta commented on src/games/strategy/common/ui/BasicGameMenuBar.java:
```diff
@@ -413,7 +413,7 @@ public void actionPerformed(final ActionEvent e) {
   protected void addAboutMenu(final JMenu parentMenu) {
     final String text = "<h2>" + getData().getGameName() + "</h2>" + "<p><b>Engine Version:</b> "
-        + games.strategy.engine.EngineVersion.VERSION.toString() + "<br><b>Game:</b> " + getData().getGameName()
+        + games.strategy.engine.ClientContext.getInstance().engineVersion().getVersion().toString() + "<br><b>Game:</b> " + getData().getGameName()
```
This update will be around a lot, changing client code to get the EngineVersion not through `EngineVersion.VERSION`, but instead `ClientContext.getIntance().engineVersion()`

ClientContext has a pattern going where the name of the object you want is the name of the method. I considered naming things as "getObject()" instead of "object()" , but in the context above, since you almost always have the "getInstance()" call, it's a whole lot of "gets" in one line. What then convinced me was nice symmetry in ClientContext from this change, things were pretty simple there and well related by names. That is important since that class will grow a lot, and if it doesn't start dead simple then we'll be in trouble.

DanVanAtta commented on src/games/strategy/engine/ClientFileSystemHelper.java:
```diff
@@ -0,0 +1,149 @@
+package games.strategy.engine;
```
This class is just pulling methods off from GameRunner2.

DanVanAtta commented on src/games/strategy/engine/EngineVersion.java:
```diff
@@ -1,7 +1,33 @@
+
+  public EngineVersion(PropertyReader propertyReader) {
+    exactVersion = propertyReader.readProperty(GameEngineProperty.ENGINE_VERSION);
+    version = new Version(exactVersion);
```
The existing code works fine to parse the 5 digit version number into a 4 digit version number. It cares if you have no ints, but it doesn't check a max, and it iterates over only the first 4 to create the `Version` number.

DanVanAtta commented on src/games/strategy/engine/EngineVersion.java:
```diff
@@ -1,7 +1,33 @@
+  private final String exactVersion;
+
+  public EngineVersion(PropertyReader propertyReader) {
+    exactVersion = propertyReader.readProperty(GameEngineProperty.ENGINE_VERSION);
```
storing original value so we can print it on the label.

DanVanAtta commented on src/games/strategy/engine/EngineVersion.java:
```diff
@@ -1,7 +1,33 @@
+  private final Version version;
+  private final String exactVersion;
+
+  public EngineVersion(PropertyReader propertyReader) {
```
PropertyReader here is dependency injection. The test case for this is very simple because we don't have to write to a file to test this. A first direction was to pass in a File to the constructor, parse it to get the EngineVersion, but that made a difficult. Either we need static initialization code somewhere (which is basically what we have in ClientContext), or each client needs to have a `PropertyReader` object it can pass in when it wants ot use an EngineVersion.

DanVanAtta commented on src/games/strategy/engine/config/GameEngineProperty.java:
```diff
@@ -0,0 +1,17 @@
+package games.strategy.engine.config;
```
Probably a nicer way to do this, #427 for suggestions on how to improve.

DanVanAtta commented on src/games/strategy/engine/config/GameEnginePropertyFileReader.java:
```diff
@@ -0,0 +1,53 @@
+
+  @Override
+  public String readProperty(GameEngineProperty propertyKey) {
+    try (FileInputStream inputStream = new FileInputStream(propertyFile)) {
```
This is code that used to be in MapDownloadController. We move it here so we do not need to duplicate it for the EngineVersion class.

DanVanAtta commented on src/games/strategy/engine/data/properties/DoubleProperty.java:
```diff
@@ -49,7 +49,7 @@ public void setValue(final Object value) throws ClassCastException {
       throw new RuntimeException(
           "Double and Number properties are no longer stored as Strings. You should delete your option cache, located at "
-              + new File(GameRunner2.getUserRootFolder(), "optionCache").toString());
+              + new File(ClientFileSystemHelper.getUserRootFolder(), "optionCache").toString());
```
Will see a number of this update, methods moved from GameRunner2 were: getUserRootFolder(), getMapsFolder().

GetFile() in GameRunner2 was moved to LoadGameUtil.

DanVanAtta commented on src/games/strategy/engine/framework/GameRunner2.java:
```diff
@@ -812,43 +810,10 @@ public void run() {
-  /**
-   * Our jar is named with engine number and we are in "old" folder.
-   */
-  public static boolean areWeOldExtraJar() {
```
Moved to ClientFileSystemHelper

DanVanAtta commented on src/games/strategy/engine/framework/GameRunner2.java:
```diff
@@ -867,147 +832,4 @@ public static Image getGameIcon(final Window frame) {
     return img;
   }
 
-  public static File getUserRootFolder() {
```
Moved to ClientFileSystemHelper

DanVanAtta commented on src/games/strategy/engine/framework/GameRunner2.java:
```diff
@@ -867,147 +832,4 @@ public static Image getGameIcon(final Window frame) {
-  /**
-   * Get the root folder for the application
-   */
-  public static File getRootFolder() {
```
Moved to ClientFileSystemHelper

DanVanAtta commented on src/games/strategy/engine/framework/GameRunner2.java:
```diff
@@ -867,147 +832,4 @@ public static Image getGameIcon(final Window frame) {
-   * @return Throws illegal state if not found. Otherwise returns a file reference whose name
-   *         matches the first parameter and parent folder matches an element of "possibleFolders"
-   */
-  public static File getFile(final String game, final String[] possibleFolders) {
```
Moved to *LoadGameUtil*  There is one method in common between the two, otherwise the next two methods also moved to that class.

DanVanAtta commented on src/games/strategy/engine/framework/mapDownload/MapDownloadController.java:
```diff
@@ -24,9 +23,8 @@
   private static final String TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES = "triplea.lastCheckForMapUpdates";
   private final MapListingSource mapDownloadProperties;
 
-  public MapDownloadController() {
```
Now the MapDownlaodController does not know anything about config file parsing : )

DanVanAtta commented on src/games/strategy/engine/framework/mapDownload/MapListingSource.java:
```diff
@@ -20,31 +14,13 @@
-  public MapListingSource(File mapDownloadPropertiesFile) {
-    checkState(checkNotNull(mapDownloadPropertiesFile).isFile());
-    mapListDownloadSite = readMapListDownloadSitePropertyValue(mapDownloadPropertiesFile);
+  public MapListingSource(PropertyReader propertyReader) {
```
Same for MapListingSource, it got a ton simpler now that it does not need to do  file parsing.

DanVanAtta commented on src/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java:
```diff
@@ -43,6 +44,9 @@
 
 public class GameSelectorPanel extends JPanel implements Observer {
   private static final long serialVersionUID = -4598107601238030020L;
+
```
Here we get to where the output labels for the engine number were added.

DanVanAtta commented on src/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java:
```diff
@@ -311,7 +343,7 @@ public void run() {
     }
     // we don't want them starting new games if we are an old jar
-    if (GameRunner2.areWeOldExtraJar()) {
+    if (ClientFileSystemHelper.areWeOldExtraJar()) {
```
From here the rest of the changes are standard method moves. Followed by that are updated test cases and new ones.

DanVanAtta commented on src/games/strategy/engine/ClientFileSystemHelper.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+//    final String tripleaJarNameWithEngineVersion = getTripleaJarWithEngineVersionStringPath();
+    if (fileName.contains("triplea_") && fileName.contains(".jar!")) {
```
EXCEPT for here -- lines 36 through 40 are new. Before we just had the call on line 35 that is now commented out. This update is necessary to allow us to read the game engine version from a configuration file.

DanVanAtta replied:
Yet another update, looks like calling 'getRootFolder()' really was just not needed, so set this back just in case..

ron-murhammer commented:
@DanVanAtta Looks like this needs rebased.

DanVanAtta commented:
merge conflict fixed

DanVanAtta commented:
@gaborbernat , @ron-murhammer , if you have any questions about the what or why's on this PR, please ask those sooner rather than later. Soon I'll lose context to these changes and will be no better off in answering those questions than either of you.

gaborbernat commented:
Review status: 0 of 59 files reviewed at latest revision, 15 unresolved discussions.

---

*[src/games/strategy/common/ui/BasicGameMenuBar.java, line 416 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1X-r1-416:-KAOrt1C6W9EVfMTsAxu:-1307484650) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/common/ui/BasicGameMenuBar.java#L416)):*
I dislike the engineVersion.getVersion(), also the getInstance(). 

getInstance() -> make it enum, and INSTANCE instead
engineVersion().getVersion() - read the engine version in the constructor of the enum, and then provide public final static fields for engineVersion/version. 

I propose this approach as the engineVersion will not change after starting up the jvm.

---

*[src/games/strategy/engine/config/GameEngineProperty.java, line 1 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKHav-09NNaMFle-r1-1:-KAOshKjG68TRXNBNGj7:965195654) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/engine/config/GameEngineProperty.java#L1)):*
yes, refactor this into an enum class with string properties, feel free to overwrite the toString too, while also making the key field public final as discussed there.

---

*[src/games/strategy/engine/EngineVersion.java, line 13 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1Z-r1-13:-KAOtk-s6f47WTPO0dpD:-691156771) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/engine/EngineVersion.java#L13)):*
given that fields are final immutable String objects I would prefer dropping the getter functions and make the public, to keep cleaner the code that uses this.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/426)*
<!-- Sent from Reviewable.io -->

gaborbernat commented:
Reviewed 53 of 60 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/426)*
<!-- Sent from Reviewable.io -->

DanVanAtta commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[src/games/strategy/common/ui/BasicGameMenuBar.java, line 416 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1X-r1-416:-KAPC61dAVuoSnAw53gt:-1223646370) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/common/ui/BasicGameMenuBar.java#L416)):*
A couple of problems that perhaps prevent the enum approach:
- the instance is not final. We need to be able to override it for test.
- initialization needs to happen when getInstance() is called, not before. This might be okay with the enum approach, but we really need to be sure about this to avoid a circular dependency with the initialization of  "GameRunner2.getRootFolder()", which only then allows us to read the game engine version.

---

*[src/games/strategy/engine/EngineVersion.java, line 13 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1Z-r1-13:-KAPELbisHFhbOMQ4HKZ:-1323473055) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/engine/EngineVersion.java#L13)):*
If there were no logic in the class EngineVersion, and it simply was  a holder for the two immutable objects, I would agree without reservation. In this case there is implementation that could change. We could for example very well hold onto a reference to "PropertyReader" and use it only when needed and return the values directly in the method calls.

Perhaps better method names would help a bit too:
-  `engineVersion.compatibilityVersion()` (rather than `engineVersion.getVersion()`)
-  `engineVersion.exactVersion()`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/426)*
<!-- Sent from Reviewable.io -->

gaborbernat commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[src/games/strategy/common/ui/BasicGameMenuBar.java, line 416 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1X-r1-416:-KAPGNPxRYhtGcXOeGGU:984884902) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/common/ui/BasicGameMenuBar.java#L416)):*
I see, still I hate how we transformed Engine.VERSION into games.strategy.engine.ClientContext.getInstance().engineVersion().getVersion().toString() ; which is way to long for my taste.

---

*[src/games/strategy/engine/EngineVersion.java, line 13 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1Z-r1-13:-KAPHFVYLYSvWG3P5HNS:408710115) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/engine/EngineVersion.java#L13)):*
I'm for the name changes. I would prefer to plan for what we have instead of what it could happen, as the YAGNI principle goes.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/426)*
<!-- Sent from Reviewable.io -->

DanVanAtta commented:
Review status: 36 of 60 files reviewed at latest revision, 14 unresolved discussions.

---

*[src/games/strategy/common/ui/BasicGameMenuBar.java, line 416 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1X-r1-416:-KAPR71YF34n1SLPgx5M:-356203894) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/common/ui/BasicGameMenuBar.java#L416)):*
It doesn't help we are missing the import statement, but the point is taken. Couple ways to improve: 
-  have 'toString()' of EngineVersion default to 'getVersion().toString()'.
-  Convert 'engineVersion()' to be static, and have `engineVersion` instead do the 'getInstance()` call

With the above it would be:
`ClientContext.engineVersion()`, rather than `Clientcontext.getInstance().engineVersion().getVersion().toString()`

---

*[src/games/strategy/engine/config/GameEngineProperty.java, line 1 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKHav-09NNaMFle-r1-1:-KARA6-HFWXUihoMn9HW:677588487) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/engine/config/GameEngineProperty.java#L1)):*
done

---

*[src/games/strategy/engine/EngineVersion.java, line 13 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1Z-r1-13:-KARA8CwA9u4VgU_lUXV:677588487) ([raw file](https://github.com/triplea-game/triplea/blob/475733b82edc6f7abd9663192f7c31a20e97cc88/src/games/strategy/engine/EngineVersion.java#L13)):*
done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/426)*
<!-- Sent from Reviewable.io -->

DanVanAtta commented:
@gaborbernat , turns out removing the getters in EngineVersion and making the immutable data members public caused an NPE in tests. I updated this PR to still get the Engine Version method rename in, as well as the simpliciation to `ClientContext.getInstance().engineVersion()` to `ClientContext.engineVersion()`

gaborbernat commented:
I like it now.

DanVanAtta commented:
This is a larger change, @ron-murhammer , @veqryn , any comments before it is merged?

ron-murhammer commented:
@DanVanAtta Let me try to find some time this evening to at least look over it at a high level. Probably won't go line by line since that appears to already be covered.

ron-murhammer commented on src/games/strategy/common/ui/BasicGameMenuBar.java:
```diff
@@ -413,7 +413,7 @@ public void actionPerformed(final ActionEvent e) {
   protected void addAboutMenu(final JMenu parentMenu) {
     final String text = "<h2>" + getData().getGameName() + "</h2>" + "<p><b>Engine Version:</b> "
-        + games.strategy.engine.EngineVersion.VERSION.toString() + "<br><b>Game:</b> " + getData().getGameName()
+        + games.strategy.engine.ClientContext.engineVersion().getCompatabilityVersion().toString() + "<br><b>Game:</b> " + getData().getGameName()
```
Is toString() needed here?
.

ron-murhammer commented on src/games/strategy/engine/config/GameEnginePropertyFileReader.java:
```diff
@@ -0,0 +1,53 @@
+        return props.getProperty(propertyKey.toString()).trim();
+      }
+    } catch (FileNotFoundException e) {
+      throw Throwables.propagate(e);
```
Not that familiar with Throwables.propagate. What is the purpose of catching then rethrowing the exception here?

ron-murhammer commented on src/games/strategy/engine/config/PropertyNotFoundException.java:
```diff
@@ -0,0 +1,9 @@
+  private static final long serialVersionUID = -7834937010739816090L;
+
+  public PropertyNotFoundException(GameEngineProperty property) {
+    super("Could not find property: " + property.toString() + ", in game engine configuration file: "+ GameEnginePropertyFileReader.getConfigFilePath());
```
Is toString() needed here?

ron-murhammer commented:
@DanVanAtta - Overall, looks pretty good and a step in the right direction.

Questions:
1. Do we need to call toString() on Version? This is done in a bunch of places, a few which I commented on above.

2. I'm not a big fan of replacing the short 'EngineVersion.VERSION' with 'ClientContext.engineVersion().getCompatabilityVersion()'. It would be nice to make this a little shorter since its used in a lot of places.

3. It looks like the 'Game Options...' button is much further down the screen now. Did you intend to move this further down the main menu or was this a side effect as I didn't see it called out?

4. So I'm a huge fan of finally displaying the engine version but my thought is it would be better to actually separate it from the 'map' info and buttons on the home screen. Everything in the left panel is related to what map you've selected and actions on that.

5. It looks like you've renamed 'game version' to 'map version' which I totally agree with. My only issue is that 'game version' is used in a few other places that weren't updated so its now kind of confusing with map vs game. An example is the Help > About popup while in game.

DanVanAtta commented:
Interesting questions, apologies for the answers being so involved.. In summary a lot of the issues we are running up are existing design flaws. For example "game" vs "engine" mean the same thing to users, but they have very distinct meanings for developers. Thus there were a number of problems to begin, I don't think we a really making any of them too much worse. The loose ends I think are probably better addressed on the spot when we see them (in the form of additional PRs)

For number 1 I fixed the places where toString() could be removed. 

For number two:
> I'm not a big fan of replacing the short 'EngineVersion.VERSION' with 'ClientContext.engineVersion().getCompatabilityVersion()'. It would be nice to make this a little shorter since its used in a lot of places.

Given that we've complicated things in two ways, it's hard to stay as clean as the enum syntax. The complications were:
- engine version is no longer a single value, but instead is a pair value
- runtime binding vs compile time

We cannot get the engine version into an enum easily since enums are statically initialized. If that happens then we are racing static initialization of reading from the local file system to then send that value into the game engine version, vs anything that happens to initialize that enum. What makes it worse, the code that reads the local file system is shared as it reads other game properties. 

Perhaps with naming we can get shorter, but it's hard to figure out how to do so. We dropped the unnecessary "get" prefix from the ClientContext methods, as well as the "getInstance()".  High consistency is being favored as well, so I wouldn't want to break the convention of "engineVersion" referring to an instance of the "EngineVersion" class, it's a useful convention. I also wouldn't want to expose all of the transitive APIs through ClientContext, which would become a god class (rather than an initialization and wiring of things together class)

"Context" refers back to a Spring "ApplicationContext", and "Client" is a bit more specific than "Game", and the 2 character difference is not worth the ambiguity. Which leaves "getCompatibilityVersion()", which could be less wordy for sure - I'm open to suggestions on that one.  At the end of the day it does not seem we can further simplify things without impacting code clarity.


> It looks like the 'Game Options...' button is much further down the screen now. Did you intend to move this further down the main menu or was this a side effect as I didn't see it called out?

Pretty much a side effect of the GridBagLayout. I didn't notice it moved down that much further down. The download maps button is landing in the left hand panel in a future PR. I would opt we fix the extra spacing after that change so we don't have to redesign it right away.

> So I'm a huge fan of finally displaying the engine version but my thought is it would be better to actually separate it from the 'map' info and buttons on the home screen. Everything in the left panel is related to what map you've selected and actions on that.

Some additional considerations I would mention:
- The game engine can change depending upon the map/save type, so in that way the two data points have a relationship. You could still think of everything as being map selection related.
- Before it was very easy to confuse "game version" to mean "engine version." In which case confused users would have already thought there was a mixture of map and game engine information on the left hand panel.
- To make the last problem worse, the UI does not make it clear that everything in the left hand panel is map selection related. I didn't realizes that myself until working closely with the code.

Thus, my response is at least it's clear we have a mixture of the game engine information with map version. Before that was not clear.

The location I do think is relatively appropriate. Maybe not the best, but it is also hard to find a better one. The current location is useful since you always want to know which engine version you are using when launching a game. For example, if it were on another panel then we would have to somehow work that other panel into the network hosting/join game screens as well. Currently the left hand panel is conveniently present on all game staging/hosting screens.

> It looks like you've renamed 'game version' to 'map version' which I totally agree with. My only issue is that 'game version' is used in a few other places that weren't updated so its now kind of confusing with map vs game. An example is the Help > About popup while in game.

This is good to know. IMO we should hunt/replace them, submit new PRs to fix. 

There is a deeper problem where the terminology of "map" vs "game" vs "engine" is not clear to the casual user. We have clear meanings for them as developers: a map contains many games and is run by an engine. But it's very confusing for users since: map, game, and engine all can pretty much mean the same thing. The fix is going to be to avoid using the word "game". "game" can sometimes mean the same thing as "map" or the "engine" from the point of view of users, but "map" and "engine" are clearly different.

A second problem is straightening out the versions. The game version refers to the version found in the XML file. I would suggest it's not a heavily used feature where people get save game compatibility thanks to one XML file having its version number bumped while the other did not. Plus the version number does not guarantee compatability anyways. I would suggest we probably should remove game versions completely in favor of just using map versions as their replacement. Then we would only need to talk about two version numbers, engine and map, and so in that setup all "games" would have the same "map" version. This is conceptually much simpler, I think particularly so since it's realy not clear where the "map version" is coming from, nor do we even display it. 

---

Review status: 36 of 59 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/games/strategy/common/ui/BasicGameMenuBar.java, line 416 \[r5\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKG2ZV6J203qg1X-r5-416:-KBGdtzJETezetXdyUqb:513233722) ([raw file](https://github.com/triplea-game/triplea/blob/3cd5473576aa1e8ffaa2b6dff39e86908eb25aa9/src/games/strategy/common/ui/BasicGameMenuBar.java#L416)):*
02e8750 removes some unnecessary toString() calls and the fully qualified package name. In some of the cases toString() is still necessary

---

*[src/games/strategy/engine/config/GameEnginePropertyFileReader.java, line 43 \[r5\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKHav-09NNaMFlf-r5-43:-KBGeGKnfFmuHZK4BN0q:459092291) ([raw file](https://github.com/triplea-game/triplea/blob/3cd5473576aa1e8ffaa2b6dff39e86908eb25aa9/src/games/strategy/engine/config/GameEnginePropertyFileReader.java#L43)):*
We've a checked exception which we are converting to an unchecked one. 

The example usage is somewhat similar: usage: http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Throwables.html#propagate(java.lang.Throwable)


We could have written instead:
```
      throw new RuntimeException(e);
```

The guava propagate has a few benefits:
 - is more explicit about what is happening. Similar reasons for using "isEmpty()" vs a check of "size() == 0".  
-  the extra RuntimeException makes the call trace a little more difficult

This post sums it up pretty well:
http://stackoverflow.com/questions/5090417/how-do-i-use-throwables-propagateifinstanceof-from-google-guava
> "The purpose of these methods is to provide a convenient way to deal with checked exceptions.
Throwables.propagate() is a shorthand for the common idiom of retrowing checked exception wrapped into unchecked one (to avoid declaring it in method's throws clause)."

This one explains what code Throwables.propagate  is replacing: http://mwenus.blogspot.com/2013/12/handling-exceptions-with-guava.html

---

*[src/games/strategy/engine/config/PropertyNotFoundException.java, line 7 \[r5\]](https://reviewable.io:443/reviews/triplea-game/triplea/426#-K9aFJKHav-09NNaMFlg-r5-7:-KBGiK-7Xh7Ad9AVmtLm:-1812290835) ([raw file](https://github.com/triplea-game/triplea/blob/3cd5473576aa1e8ffaa2b6dff39e86908eb25aa9/src/games/strategy/engine/config/PropertyNotFoundException.java#L7)):*
Fixed in : 02e8750

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/triplea-game/triplea/426#-:-KBGuVeTSwWXohGfmWne:264401341)*
<!-- Sent from Reviewable.io -->

ron-murhammer commented:
@DanVanAtta - Alright. I'm fine holding off on some of the things mentioned above. The only one that would probably be good to address in this PR is replacing the short 'EngineVersion.VERSION' with 'ClientContext.engineVersion().getCompatabilityVersion()'. I understand the rework so we no longer have an enum or final static value but would like to shorten the expression. Here are some suggestions that I think would be better:

1. 'ClientContext.engineVersion().getVersion()' or 'ClientContext.engineVersion().toString()' - Just consider it 'version' or 'default toString' since it is the one we primarily use. Could keep exactVersion or rename to something like fullVersion or bulidVersion. 

2. 'ClientContext.engineVersion()' - merge EngineVersion functionality into Version so that this returns Version and EngineVersion class goes away. Essentially this would just be adding 5 digit version number instead of 4 digit version number support into the version class.

3. 'ClientContext.getCompatabilityVersion()' - could just merge EngineVersion into ClientContext since both are so small and have very limited functionality so the versions could be returned directly.

I'm open to any of the above or combinations of them. Really just would like to have a shorter expression for getting the core engine version than 'ClientContext.engineVersion().getCompatabilityVersion()'.

DanVanAtta commented:
Two changes added to help:
1. shorten/rename:
- getExactVersion() -> getFullVersion()
- getCompabilityVersion() -> getVersion()
2. Forward calls to `EngineVersion.toString()` to `EngineVersion.getVersion().toString()`

Below are some below and after counts, so those two changes helped:

```
## Before
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion().getCompatabilityVersion().toString()" {} \; | wc -l
8
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion().getCompatabilityVersion()" {} \; | wc -l
44
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion().getExactVersion()" {} \; | wc -l
1
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion().toString()" {} \; | wc -l
0
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion()" {} \; | wc -l
48


## After
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion().getVersion()" {} \; | wc -l
28
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion().getFullVersion()" {} \; | wc -l
2
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion().toString()" {} \; | wc -l
8
dan@dan-desk:~/work/triplea$ find . -name "*java" -exec grep "ClientContext.engineVersion()" {} \; | wc -l
48
```

DanVanAtta commented:
In summary @ron-murhammer , this is as far as I think we can go for now without adding in an IOC  framework.  We really do not want to add any implementation to ClientContext. I see ClientContext as either growing to have many classes or being replaced wholesale by either XML or annotations (IOC frameworks like Spring/Dagger would allow that)


For a much longer answer, please read further (maybe find a comfortable chair as well). Perhaps most of the rest you can skim, but please do pay some attention to the last two code examples. IOC can quickly turn into Global Variables masquarading as IOC when not done well. 


Beyond the changes already applied, I do not think we can go any further. It comes down to the question of do we force the following construct:
- `A.B().C()`  + `A.B().D()`
Or do we allow:
- `A.C()` + `A.D()`


If we want to acheive the second, where we get the value directly from A, there are two ways to do this:
- ClientContext acts like a facade class and calls to `C()` and `D()` are passed to `B()` "under the covers"
- Alternatively we break down `B()` and move it into A. In our case that would mean deleting `EngineVersion.java` and adding two variables to `ClientContext`: 
  - `private Version engineVersion`
  - `private String fullVersion`


The problem with the second approach is that we've added implementation in ClientContext. There is no limit really on the objects that ClientContext would manage, so it's truly killer ot have implementation. Even the facade implementation is killer. Let's say ClientContext were to manage "n" objects, each with "m" methods. If it were a facade then we would have "n*m-n)" public APIs, but if we have one public method per class than we would have only "n".

I think this fundamentally gets to the crux and future of ClientContext. ClientContext is a POJO inversion of control container. Either it'll grow and will pick up more and more classes that it will manage, one day perhaps hitting 20, 50, and beyond.. Or alternatively we will replace it with Spring or Dagger. In the latter case we would get rid of ClientContext completely and would sprink annotations on the classes that ClientContext would otherwise manage. 

The last scenario where we adopt Spring or Dagger in my mind is probably the best, and it achieves the simplification we would want. For example instead of:

```
public Foo( ) {
  this(ClientContext.sharedInitializionObject())
}

public Foo( Bar dependency) {
  dependency.toString()
}
```

With dagger for examle we would just have:
```
@Injected
public Foo( Bar dependency) {
  dependency.toString()
}
```


So the final part to this answer - it's a code smell for us to have a lot of:
`ClientContext.static`

If we do a lot of that then we don't have dependency injection, we have global variables instead. So partly what has happened here is a best effort attempt at a clean'ish migration. More ideally dependencies would be injected as constructor args to the classes that need them. Given the existing structure, that would be difficult to do in many places.

So, as an example, this is not good:
```
public class myClass() {
	 public myClass() {

	 }
	 public void foo() {
	    DependencyObject obj = ClientContext.mayAsWellBeGlobalVariable();
	    obj.doStuff();
	 }
}
```

This is what we should shoot for instead:

```
public class myClass() {
    private final DependencyObject obj;
	 public myClass(final DependencyObject obj) {
	    this.obj = obj;
	 }
	 public void foo() {
	    obj.doStuff();
	 }
}
```

So in this PR we certainly have a lot of the first example. It would be a massive effort to get the existing classes into the second structure. But, we now have some basic code infrastructure that could help us peel away functionality into classes that look like the second example, and then are wired together via dependency injection. The classes that are wired together are always going to be loosely coupled. The system today since most classes initialize their dependencies, we have very strong coupling between those classes and their dependencies (which makes dependency sharing nasty, often we've formed inappropriate inheritance heirarchies to solve this problem rather than using IOC).

ron-murhammer commented:
@DanVanAtta Agree. I'm good with the changes and that was what I figured at this point is the best solution. 

In regards to IOC, it definitely would be a challenge at this point and probably not very beneficial given the way most of the code is structured. This also leads to a lot of the challenges around testing right now as things are so tightly coupled. 

It would be nice to eventually get there though I think the code base as a whole needs restructured and broken down into separate projects and then many classes need refactored to be more modular. I think those are probably the things we'll more likely tackle first.

DanVanAtta commented:
> "It would be nice to eventually get there though I think the code base as a whole needs restructured and broken down into separate projects and then many classes need refactored to be more modular. I think those are probably the things we'll more likely tackle first."

I'm in more agreement than I think you perhaps realize.


For example, let's say we have this:

```
public class TripleaGodClass {
  public TripleaGodClass() {
        LogicClass tightCoupling = new LogicClass();
        YouCantTestMe hardToTest = new YouCantTestMe();
      :
      :
  }
  :
 :
}
```

If we notice that the `LogicClass` maybe also needs to be used somewhere, or is a good thing to extract, IOC can help:

```
public class TripleaGodClass {
  public TripleaGodClass(LogicClass canBeMockedNow) {
        looseCoupling = canBeMockedNow();
        YouCantTestMe hardToTest = new YouCantTestMe();
      :
      :
  }
  :
 :
}
```
I see this as one path towards untangling things.  Or perhaps as a better example:

```
public class TripleaComplicatedClass {
  public TripleaComplicatedClass( ) {
     // logic
     // detailed logic
     // logic
     // ultra detailed logic
      :
      :
  }
  :
 :
}
```

To:

```
public class LogicClass() {
     // detailed logic
     // ultra detailed logic
}

public class TripleaComplicatedClass {
  public TripleaComplicatedClass(LogicClass logicClass ) {
     // logic
     // logic
      logicClass.handleDetails( logic) 
      :
      :
  }
  :
 :
}
```

DanVanAtta commented:
Thank you @gaborbernat  and @ron-murhammer  for the review/feedback
